joomla/joomla-cms
# [4.0] poor man’s cron

porting #22310 to 4.x

### Goal
a way to run scheduled "quick" tasks in approximately regular intervals

### Summary of Changes
- a new system plugin scheduler
- two new console command 

### How it works
In a normal hosting situation that would be job for cron, but if your hosting doesn’t offer cron, then a common solution in these cases is “poor man’s cron:” on every page load (`onAfterRespond`) the scheduler system plugin check if itself  is expected to run and  then trigger a new event `onExecuteScheduledTask`  then any job plugin wich listen to that event  check if itself  is expected to run and if so execute his task.  

### Testing Instructions (web)

- apply PR 
- System panel -> discover -> you'll find a new system plugin scheduler
- install it 
- set options, publish the plugin and save
for test only
Frequency (in minutes) set = 0 mean run every (`onAfterRespond`)

![Screenshot from 2019-05-28 18-55-57](https://user-images.githubusercontent.com/181681/58496527-5e52f900-817a-11e9-9aec-3e96347fd91c.png)

- install these 2 silly job plugin:
[plg_job_one.zip](https://github.com/joomla/joomla-cms/files/3238278/plg_job_one.zip)
[plg_job_two.zip](https://github.com/joomla/joomla-cms/files/3238272/plg_job_two.zip)


-  job one simply sleep for 1 seconds		
-  job two simply sleep for 3 seconds

- set options, publish the plugin and save plugin ONE
Frequency (in unit) set 1
Unit of time set minute

 means that this plugin run 1xminute = every minute 
if `onExecuteScheduledTask`  has been triggered 

![Screenshot from 2019-05-30 18-55-18](https://user-images.githubusercontent.com/181681/58649498-896d5200-830c-11e9-823f-40c06f2d31f2.png)

- set options, publish the plugin and save plugin TWO
maybe use different  settings than ONE

### Expected result
depending on your (plugins) settings 
take a coffe/smoke/whatever :smile: 
randomly click a couple of  frontend pages
check the log file at  `yoursite/administrator/logs/joomla_web.php`
you'll see something like this 
```php
#
#<?php die('Forbidden.'); ?>
#Date: 2019-05-30 17:24:39 UTC
#Software: Joomla! 4.0.0-alpha10-dev Development [ Amani ] 16-May-2019 19:22 GMT

#Fields: datetime	priority clientip	category	message
2019-05-30T17:24:39+00:00	INFO 127.0.0.1	scheduler	Scheduler>start
2019-05-30T17:24:40+00:00	INFO 127.0.0.1	scheduler	Job>jobone Task#221 Processing Time: 1.01 seconds.
2019-05-30T17:24:43+00:00	INFO 127.0.0.1	scheduler	Job>jobtwo Task#166 Processing Time: 3.01 seconds.
2019-05-30T17:24:43+00:00	INFO 127.0.0.1	scheduler	Scheduler>stop>task#1564 Total Processing Time: 4.04 seconds.
2019-05-30T17:24:47+00:00	INFO 127.0.0.1	scheduler	Scheduler>start
2019-05-30T17:24:47+00:00	INFO 127.0.0.1	scheduler	Scheduler>stop>task#1565 Total Processing Time: 0.04 seconds.
```

 **1st run** 
the scheduler plugin (wich runs always frequency=0) trigger the `onExecuteScheduledTask` event
so plugin ONE and plugin TWO  are executed (cause no previous run)

**2nd  run**
the scheduler plugin (wich runs always frequency=0) trigger the `onExecuteScheduledTask` event
but this time both plugin have an expected scheduled time (for example ONE each 1min TWO each 5min) 
the 2nd run was at 17:24:47+00:00	
the 1st run was at 17:24:43+00:00
they have only 4 seconds of difference form the previous execution so both plugin are not expected to run

**3rd  run** 
it depends on how fast you click & settings


### Testing Instructions (webcron)
on scheduler system plugin
set Webcron to yes/enable 
set Key to mywebcronactivationkey
then use `yoursiteurl/index.php?webcronkey=mywebcronactivationkey` as the url for webcron services to be triggred


### Testing Instructions (cli/cron)
Go to the cli folder of your website and run
`php joomla.php job:run`
![Screenshot from 2019-05-30 20-03-51](https://user-images.githubusercontent.com/181681/58653556-3f896980-8316-11e9-89aa-f8c3607ae779.png)

check the log file at yoursite/administrator/logs/joomla_cli.php


### Actual result
N/A

### Additional comments

it can be used for tasks like:

- deleteExpiredConsents
- remindExpiringConsents
- logRotation
- updateNotifications
- sessionGC
- etc...

### Documentation Changes Required
yes
